### PR TITLE
record the use of entry field

### PR DESCRIPTION
**Pull request by chaychen2005**

### CODE REVIEW

File: libstorage/BinLogHandler.cpp [hunk 1]
```diff
@@ -394,21 +413,39 @@ uint32_t BinLogHandler::decodeEntries(const bytes& buffer, uint32_t& offset,
         /*BINLOG_HANDLER_LOG(TRACE) << LOG_DESC("entry info") << LOG_KV("idx in entries", idx)
                                   << LOG_KV("_id_", entry->getID())
                                   << LOG_KV("_status_", (uint32_t)entry->getStatus());*/
+        int flagSize = (vecField.size() - 3) / 8 + 1;  // exclude STATUS,NUM_FIELD,ID_FIELD
+        uint8_t* usedFlag = new uint8_t[flagSize];     // bit-based used to represent a field
+        int fieldIdx = 0;                              // field index
+        for (int i = 0; i < flagSize; i++)
```
Comment (bxq2011hust) [inline libstorage/BinLogHandler.cpp:419]:
```suggestion
       uint8_t * usedFlag = buffer+ offset;
       offset+= flagSize;
```

File: libstorage/BinLogHandler.cpp [hunk 2]
```diff
@@ -304,19 +304,39 @@ void BinLogHandler::encodeEntries(
                                   << LOG_KV("id", entry->getID())
                                   << LOG_KV("status", (uint32_t)status);*/
         buffer.insert(buffer.end(), (byte*)&status, (byte*)&status + 1);
+        int flagSize = (vecField.size() - 3) / 8 + 1;  // exclude STATUS,NUM_FIELD,ID_FIELD
+        uint8_t* usedFlag = new uint8_t[flagSize];     // bit-based used to represent a field
+        memset(usedFlag, 0, sizeof(uint8_t) * flagSize);
+        int KeyIdx = 0;  // field index
+        for (const auto& key : vecField)
+        {
+            if (key == STATUS || key == NUM_FIELD || key == ID_FIELD)
```
Comment (bxq2011hust) [inline libstorage/BinLogHandler.cpp:313]:
```suggestion
            if (key == STATUS || key == NUM_FIELD || key == ID_FIELD)
```
This filiter can be done of vecField

File: libstorage/BinLogHandler.cpp [hunk 3]
```diff
@@ -304,19 +304,39 @@ void BinLogHandler::encodeEntries(
                                   << LOG_KV("id", entry->getID())
                                   << LOG_KV("status", (uint32_t)status);*/
         buffer.insert(buffer.end(), (byte*)&status, (byte*)&status + 1);
+        int flagSize = (vecField.size() - 3) / 8 + 1;  // exclude STATUS,NUM_FIELD,ID_FIELD
```
Comment (bxq2011hust) [inline libstorage/BinLogHandler.cpp:307]: binlog version should be changed?

File: libstorage/BinLogHandler.cpp (hunk 2 above)
Comment (chaychen2005) [reply libstorage/BinLogHandler.cpp:313]: these three fields recorded are required when restoring table fields

File: libstorage/BinLogHandler.cpp (hunk 3 above)
Comment (chaychen2005) [reply libstorage/BinLogHandler.cpp:307]: binlog hasn't been released yet, so don't need to change the version number

File: libstorage/BinLogHandler.cpp (hunk 1 above)
Comment (chaychen2005) [reply libstorage/BinLogHandler.cpp]: Done

File: libstorage/BinLogHandler.cpp (hunk 3 above)
Comment (chaychen2005) [reply libstorage/BinLogHandler.cpp:307]: binlog hasn't been released yet, so don't need to change the version number

File: libstorage/BinLogHandler.cpp (hunk 2 above)
Comment (chaychen2005) [reply libstorage/BinLogHandler.cpp:313]: these three fields recorded are required when restoring table fields

File: libstorage/BinLogHandler.cpp (hunk 1 above)
Comment (chaychen2005) [reply libstorage/BinLogHandler.cpp:419]: use array as a unit to process, easy to understand

File: libstorage/BinLogHandler.cpp [hunk 4]
```diff
@@ -304,19 +304,39 @@ void BinLogHandler::encodeEntries(
...
+        {
+            if (key == STATUS || key == NUM_FIELD || key == ID_FIELD)
+            {
+                continue;
+            }
+            if (entry->find(key) != entry->end())
+            {
+                // "KeyIdx / 8" get the index in array, "7 - KeyIdx % 8" get the bit index in byte
+                // "A += std::pow(2, B)" set the Bth digit of A to 1, count backwards
+                usedFlag[KeyIdx / 8] += std::pow(2, (7 - KeyIdx % 8));
```
Comment (cyjseagull) [inline libstorage/BinLogHandler.cpp:321]:
```suggestion
               usedFlag[KeyIdx / 8] |= (1 <<  (7 - KeyIdx % 8));
```

Comment (chaychen2005) [reply libstorage/BinLogHandler.cpp]: Good, done

**Pull request merged**